models/inventory/StockTransfer.ts

```diff
@@ -373,6 +421,64 @@ export abstract class StockTransfer extends Transfer {
 
     return invoice;
   }
+
+  async getReturnDoc(): Promise<StockTransfer | null> {
+    if (!this.items?.length) {
+      return null;
+    }
+
+    const docData = this.getValidDict(true, true);
+    const docItems = docData.items as StockTransferItem[];
+    const returnDocItems: StockTransferItem[] = [];
+
+    const returnBalanceItemsQty = await this.fyo.db.getReturnBalanceItemsQty(
+      this.schema.name,
+      this.name!
+    );
+
+    for (const item of docItems) {
+      if (!item.quantity) {
+        continue;
+      }
+
+      let quantity = -1 * item.quantity;
+
+      if (returnBalanceItemsQty) {
+        const balanceItemQty = returnBalanceItemsQty.filter(
+          (i) => i.item === item.item
+        )[0];
+
+        if (!balanceItemQty) {
+          continue;
+        }
+        quantity = balanceItemQty.quantity as number;
+      }
+
+      item.quantity = safeParseFloat(quantity);
+      delete item.name;
+      returnDocItems.push(item);
+    }
+
+    const returnDocData = {
+      ...docData,
+      name: null,
+      date: new Date(),
+      items: returnDocItems,
+      isReturn: true,
+      returnAgainst: docData.name,
+      grandTotal: this.fyo.pesa(0),
+    };
+
+    const rawReturnDoc = this.fyo.doc.getNewDoc(
+      this.schema.name,
+      returnDocData
+    ) as StockTransfer;
+
+    rawReturnDoc.once('beforeSync', async () => {
+      await rawReturnDoc.runFormulas();
```

Comment:
Why not just run it once before returning it, instead of calling it in `beforeSync`?

Reply:
corrected